Cql/Cql.Compiler/ExpressionBuilderContext.cs

```diff
@@ -1153,56 +1152,46 @@ protected Expression ParameterRef(ParameterRef op)
-        var selected = _libraryContext.LibraryDefinitions.Resolve(libraryName, name, CheckConversion, argumentTypes);
-        Type definitionType = GetFuncType(selected.Parameters.Select(p => p.Type).Append(selected.ReturnType).ToArray());
-        var parameterTypes = selected.Parameters.Skip(1).Select(p => p.Type).ToArray();
-
-        // all functions still take the bundle and context parameters, plus whatver the operands
-        // to the actual function are.
+        var rtt = TypeFor(returnType) ?? throw this.NewExpressionBuildingException($"Unable to resolve type for {returnType}");
         var convertedArguments = arguments
-                                 .Select((arg, i) =>
-                                 {
-                                     var argConversion = TypeConversion.NoMatch;
-                                     var converted = ChangeType(arg, parameterTypes[i], out argConversion);
-                                     if (argConversion == TypeConversion.NoMatch)
-                                         throw this.NewExpressionBuildingException($"Cannot convert {arg.Type} to {parameterTypes[i]}.");
-                                     else return converted;
-                                 })
+                                 .Select((a,i) => convertChoice(a,signature?[i]))
                                  .Prepend(CqlExpressions.ParameterExpression)
                                  .ToArray();
+        var funcType = convertedArguments.Select(a => a.Type).Append(rtt).ToArray();
+        Type definitionType = Expression.GetFuncType(funcType);
 
         return new FunctionCallExpression(CqlExpressions.Definitions_PropertyExpression, libraryName, name, convertedArguments, definitionType);
 
-        bool CheckConversion(Type from, Type to) => _typeConverter.CanConvert(from, to);
+        // This function will handle the cases where the normal C# invocation is insufficient to represent the CQL function call:
+        // the argument is of a choice type, and the parameter is of a specific type (or for now, also a choice type).
+        // In this case we need to insert a conversion from the choice type to the specific type of the argument. Presumably, the
+        // cql2elm compiler has already checked that the call is valid, but we do need to cast the choice type (in C# represented by
+        // object/DataType) to the actual type to make this a valid C# call. CQL semantics state that the result may be null if the
+        // choice is not compatible with the parameter, so we'll use an As in C#.
+        Expression convertChoice(Expression argument, TypeSpecifier? actualType)
+        {
+            if(actualType is not null && argument.Type == typeof(object))
+            {
+                var type = TypeFor(actualType) ?? throw new InvalidOperationException($"Unable to resolve type for {actualType}");
+                return argument.NewTypeAsExpression(type);
```

Comment:
Should we not do a hard cast here, otherwise a null reference/argument exception can manifest later

Reply:
We could use the ChangeType method in the other https://github.com/FirelyTeam/firely-cql-sdk/pull/496. I'll create an issue with tasks to be done after this PR is merged, from the cherry picking commits on branch 3, and including this

Reply:
> Should we not generate a hard cast here, otherwise a null reference/argument exception can manifest later.

I thought I remembered that the semantics of CQL dictate that this should be a null - there are many cases in the CQL where a "failure" turns into null/empty. But it's good to double check with @EvanMachusak .

Reply:
In that case, let's change this to the ChangeType method that was just merged. It uses safe casting.

Reply:
I will resolve this here and do it in issue #503